clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,143 @@
+use rustc_span::Span;
+use rustc_span::def_id::LocalDefId;
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for test functions (functions annotated with `#[test]`) that are prefixed
+    /// with `test_` which is redundant.
+    ///
+    /// ### Why is this bad?
+    /// This is redundant because test functions are already annotated with `#[test]`.
+    /// Moreover, it clutters the output of `cargo test` since test functions are expanded as
+    /// `module::tests::test_use_case` in the output. Without the redundant prefix, the output
+    /// becomes `module::tests::use_case`, which is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #[test]
+    /// fn test_use_case() {
+    ///     // test code
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn use_case() {
+    ///    // test code
+    /// }
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub REDUNDANT_TEST_PREFIX,
+    pedantic,
+    "redundant `test_` prefix in test function name"
+}
+
+pub struct RedundantTestPrefix {
+    check_outside_test_cfg: bool,
+    custom_sufix: String,
+}
+
+impl RedundantTestPrefix {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self {
+            check_outside_test_cfg: conf.redundant_test_prefix_check_outside_cfg_test,
+            custom_sufix: conf.redundant_test_prefix_custom_suffix.clone(),
+        }
+    }
+}
+
+impl_lint_pass!(RedundantTestPrefix => [REDUNDANT_TEST_PREFIX]);
+
+impl<'tcx> LateLintPass<'tcx> for RedundantTestPrefix {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: FnKind<'_>,
+        _decl: &FnDecl<'_>,
+        body: &'tcx Body<'_>,
+        _span: Span,
+        fn_def_id: LocalDefId,
+    ) {
+        // Ignore methods and closures.
+        let FnKind::ItemFn(ref ident, ..) = kind else {
+            return;
+        };
+
+        // Skip the lint if the function is not within a node marked with `#[cfg(test)]` attribute.
+        // Continue if the function is inside the node marked with `#[cfg(test)]` or lint is enforced
+        // via configuration (most likely to include integration tests in lint's scope).
+        if !(self.check_outside_test_cfg || is_in_cfg_test(cx.tcx, body.id().hir_id)) {
+            return;
+        }
+
+        if is_test_function(cx.tcx, cx.tcx.local_def_id_to_hir_id(fn_def_id)) && ident.as_str().starts_with("test_") {
+            let mut non_prefixed = ident.as_str().trim_start_matches("test_").to_string();
+            let mut help_msg = "consider removing the `test_` prefix";
+            // If `non_prefixed` conflicts with another function in the same module/scope,
+            // add extra suffix to avoid name conflict.
+            if name_conflicts(cx, body, &non_prefixed) {
+                non_prefixed.push_str(&self.custom_sufix);
+                help_msg = "consider removing the `test_` prefix (suffix avoids name conflict)";
```

Comment:
As mentioned at the place with the configuration, I think it would be better to just change the `Applicability` instead of adding a predefined suffix. We could still suggest renaming it to `non_prefixed.push_str("_works")` but have it as `Applicability::HasPlaceholders`